### PR TITLE
Feat/gradient

### PR DESCRIPTION
**Pull request by wjian23**

### CODE REVIEW

File: painter/src/painter.rs [hunk 1]
```diff
@@ -289,7 +327,24 @@ impl Painter {
           color: color.apply_alpha(opacity),
         },
         Brush::Image(img) => PaintCommand::ImgPath { path, img, opacity },
-        Brush::Gradient => todo!(),
+        Brush::RadialGradient(radial_gradient) => PaintCommand::RadialGradient {
+          path,
+          stops: radial_gradient.stops,
```
Comment (M-Adoo) [inline painter/src/painter.rs:332]: PaintCommand::RadialGradient  should be   ```PaintCommand::RadialGradient {
path，gradinet
}``` ?

File: gpu/src/lib.rs [hunk 2]
```diff
@@ -136,6 +172,74 @@ pub struct ColorAttr {
   pub mask_head: i32,
 }
 
+#[repr(packed)]
+#[derive(AsBytes, PartialEq, Clone, Copy, Debug)]
+pub struct RadialGradientAttr {
```
Comment (M-Adoo) [inline gpu/src/lib.rs:177]: `XXXPrimIndex(u32) ` is a better name ?

Comment (M-Adoo) [reply gpu/src/lib.rs:177]: if gradient primitive index have its own type, the primitive index of color and image should have its own type too

File: gpu/src/gpu_backend.rs [hunk 3]
```diff
@@ -8,14 +12,21 @@ use ribir_painter::{
...
   color_vertices_buffer: VertexBuffers<ColorAttr>,
-  img_vertices_buffer: VertexBuffers<u32>,
+  img_vertices_buffer: VertexBuffers<ImagePrimIndex>,
+  radial_gradient_vertices_buffer: VertexBuffers<RadialGradientPrimIndex>,
+  radial_gradient_stops: Vec<GradientStopPrimitive>,
+  radial_gradient_prims: Vec<RadialGradientPrimitive>,
+  linear_gradient_prims: Vec<LinearGradientPrimitive>,
+  linear_gradient_stops: Vec<GradientStopPrimitive>,
+  linear_gradient_vertices_buffer: VertexBuffers<LinearGradientPrimIndex>,
   img_prims: Vec<ImgPrimitive>,
```
Comment (M-Adoo) [inline gpu/src/gpu_backend.rs:30]: `img_prims` should next to `img_vertices_buffer`.

File: text/src/svg_font_document.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
+use log::warn;
+use quick_xml::events::attributes::Attribute;
+use quick_xml::events::{BytesStart, Event};
+use quick_xml::name::QName;
+use quick_xml::reader::Reader;
+use rustybuzz::ttf_parser::GlyphId;
+
+use std::borrow::Cow;
+use std::collections::HashMap;
```
Comment (M-Adoo) [inline text/src/svg_font_document.rs:9]: we use ahash::Hashmap  whole project.

File: painter/src/svg.rs [hunk 5]
```diff
@@ -67,26 +71,26 @@ impl Svg {
...
               };
 
-              let brush = brush_from_usvg_paint(&stroke.paint, stroke.opacity);
-              paths.push(SvgPath {
-                path,
-                brush,
-                style: PathPaintStyle::Stroke(options),
-              });
+              let (brush, transform) = brush_from_usvg_paint(&stroke.paint, stroke.opacity, &size);
+              let mut painter = painter.save_guard();
```
Comment (M-Adoo) [inline painter/src/svg.rs:81]: chaining methods to set painter config.

**Review by M-Adoo — APPROVED** [review]

**Review by sologeek — APPROVED** [review]

**Pull request merged**